solidity/contracts/middleware/libs/InterchainAccountMessage.sol

```diff
@@ -107,27 +117,84 @@ library InterchainAccountMessage {
+        bytes memory prefix = abi.encodePacked(
+            MessageType.CALLS,
+            _owner,
+            _ism,
+            _userSalt
+        );
+        bytes memory suffix = abi.encode(_calls);
+        return bytes.concat(prefix, suffix);
     }
 
     /**
-     * @notice Parses and returns the calls from the provided message
-     * @param _message The interchain account message
-     * @return The array of calls
+     * @notice Returns formatted (packed) InterchainAccountMessage
+     * @dev This function should only be used in memory message construction.
+     * @param _owner The owner of the interchain account
+     * @param _ism The address of the remote ISM
+     * @return Formatted message body
      */
-    function decode(
-        bytes calldata _message
-    ) internal pure returns (bytes32, bytes32, CallLib.Call[] memory, bytes32) {
+    function encodeCommitment(
+        bytes32 _owner,
+        bytes32 _ism,
+        bytes32 _commitment,
+        bytes32 _userSalt
+    ) internal pure returns (bytes memory) {
         return
-            abi.decode(_message, (bytes32, bytes32, CallLib.Call[], bytes32));
+            abi.encodePacked(
+                MessageType.COMMITMENT,
+                _owner,
+                _ism,
+                _userSalt,
+                _commitment
+            );
+    }
+
+    function encodeReveal(
+        bytes32 _ism,
+        bytes32 _commitment
+    ) internal pure returns (bytes memory) {
+        return abi.encodePacked(MessageType.REVEAL, _ism, _commitment);
+    }
+
+    function messageType(
+        bytes calldata _message
+    ) internal pure returns (MessageType) {
+        return MessageType(uint8(bytes1(_message[0:1])));
+    }
+
+    function owner(bytes calldata _message) internal pure returns (bytes32) {
+        return bytes32(_message[1:33]);
     }
 
     /**
      * @notice Parses and returns the ISM address from the provided message
      * @param _message The interchain account message
      * @return The ISM encoded in the message
      */
-    function ism(bytes calldata _message) internal pure returns (address) {
-        return address(bytes20(_message[44:64]));
+    function ism(bytes calldata _message) internal pure returns (bytes32) {
+        if (messageType(_message) == MessageType.REVEAL) {
+            return bytes32(_message[1:33]);
+        }
+        return bytes32(_message[33:65]);
+    }
+
+    function salt(bytes calldata _message) internal pure returns (bytes32) {
+        return bytes32(_message[65:97]);
+    }
+
+    function calls(
+        bytes calldata _message
+    ) internal pure returns (CallLib.Call[] memory) {
+        return abi.decode(_message[97:], (CallLib.Call[]));
+    }
+
+    function commitment(
+        bytes calldata _message
+    ) internal pure returns (bytes32) {
+        if (messageType(_message) == MessageType.REVEAL) {
+            return bytes32(_message[33:65]);
+        }
+        return bytes32(_message[97:]);
```

Comment:
when type == `CALLS`, this is invalid yes?

Reply:
Yeah, it would just be the first 32 bytes of the encoded `Call[] calls`.

Reply:
I would maybe assert against that then
what do you think of having separate libraries for each message type as opposed to branching within each function?

Reply:
Looks like we'll just end up moving to structs (https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/6166). I've already added a library for reveal messages, but I'll skip adding another lib for commitment messages.